rpmlint/checks/FileDigestCheck.py

```diff
@@ -2,23 +2,141 @@
+    contents as is."""
+
+    def __init__(self, path, halg):
+        self.path = path
+        self.halg = halg
+
+    def parse_content(self):
+        # simply generate chunks on binary level
+        with open(self.path, 'rb') as fd:
+            while True:
+                chunk = fd.read(4096)
+                if not chunk:
+                    break
+                yield chunk
+
+    def get_digest(self):
+        """This returns the hash digest of the *un*filtered input file."""
+        hasher = hashlib.new(self.halg)
+
+        for chunk in self.parse_content():
+            hasher.update(chunk)
+
+        return hasher.hexdigest()
+
+
+class ShellDigester(DefaultDigester):
+    """This class performs digest calculation of shell style configuration
+    files or scripts. Certain aspects like comments and whitespace will be
+    filtered out of the digest calculation."""
+
+    def parse_content(self):
+        # generate filtered lines
+        with open(self.path) as fd:
+            for line_nr, line in enumerate(fd):
+                stripped = line.strip()
+                if not stripped:
+                    # skip empty lines
+                    continue
+                elif line_nr == 0 and stripped.startswith('#!'):
+                    # keep shebang lines instact
+                    pass
+                elif stripped.startswith('#'):
+                    # skip comments
+                    # NOTE: we don't strip trailing comments like in
+                    # 'if [ 5 -eq $NUM ]; then # compare to 5'
+                    # because the danger would be too high to remove actual
+                    # content instead of just comments
+                    continue
+
+                # don't use the completely stripped version, in Python for
+                # example the indentation is part of the syntax and changes
+                # here could change the meaning
+                yield (line.rstrip() + '\n').encode()
+
+
+class XmlDigester(DefaultDigester):
+    """This class performs digest calculation of XML configuration files.
+    Certain aspects like comments and whitespace will be filtered out of the
+    digest calculation."""
+
+    def parse_content(self):
+        # NOTE: the ElementTree is not robust against malicious input. Rpmlint
+        # processes a lot of untrusted input. In the OBS context this only
+        # happens within the virtual machines where packagers can manipulate a
+        # lot of things anyway so it should be okay.
+
+        # this returns a canonicalized form of the XML without comments or
+        # anything. It is missing the XML preamble and DOCTYPE declaration,
+        # though. It's as good as I could get it to be without parsing XML on
+        # foot or relying on third party XML modules.
+
+        # chunked / line wise processing is likely impossible for XML so
+        # return the whole bunch
+        yield ET.canonicalize(from_file=self.path, strip_text=True).encode()
```

Comment:
Oh, CI shows that this won't work on Leap:
`E       AttributeError: module 'xml.etree.ElementTree' has no attribute 'canonicalize'`

It's pretty new (`3.8`):
https://docs.python.org/3/library/xml.etree.elementtree.html#xml.etree.ElementTree.canonicalize

Reply:
Yes I saw that before ... does it need to work on Leap?

Reply:
> Are there any plans currently to backport rpmlint2 to Leap?

No, I actually refused to update rpmlint to version 2 in 15.4. That said, I think we can request version `3.8` and later. I'm going to discuss that in another pull request. For now, we can merge your pull request as it's only for the `opensuse` branch.